### PR TITLE
Implement sticky date separators

### PR DESCRIPTION
**Pull request by lukebarnard1**
Use `react-sticky` to implement sticky date separators. This will pin a date separator to the top of the timeline panel when the separator scrolls out of the top of the view.

A known issue of this is that the spinner, which is in line with event tiles in the timeline, will appear to push the stuck date separator down. In reality the first date separator after the spinner is in line with event tiles and is not stuck because the spinner forces the timeline to be scrolled slightly further down than it would be otherwise. But also, date separators in the timeline (not "stuck") have a greater height.

Ideally the date separator would be suppressed whilst back paginating, but this will cause the stuck separator to flicker on and off. This is why the suppression has been removed.

Also, `DateUtils.formatDateSeparator` was factored out of `DateSeparator` and fixed so that it will now display:
 - Today
 - Yesterday
 - Wednesday
 - Tuesday
 - …
 - Sat, Aug 26
 - …
 - Friday, Feb 23 2013

codep vector-im/riot-web#4939

### CODE REVIEW

File: src/components/structures/ScrollPanel.js [hunk 1]
```diff
@@ -77,111 +78,48 @@ if (DEBUG_SCROLL) {
...
-    },
+    }
 
-    onScroll: function(ev) {
+    onScroll(ev) {
         var sn = this._getScrollNode();
         debuglog("Scroll event: offset now:", sn.scrollTop,
                  "_lastSetScroll:", this._lastSetScroll);
 
+        this.node = sn;
```
Comment (dbkr) [inline src/components/structures/ScrollPanel.js]: unused?

Comment (lukebarnard1) [reply src/components/structures/ScrollPanel.js]: Actually no, this is why any of the sticky stuff works and I 100% should have commented as such.

File: src/DateUtils.js [hunk 2]
```diff
@@ -96,6 +108,38 @@ module.exports = {
...
+        } else {
+            return _t('%(weekDayName)s, %(monthName)s %(day)s %(fullYear)s', {
+                weekDayName: days[date.getDay()],
+                monthName: months[date.getMonth()],
+                day: date.getDate(),
+                fullYear: date.getFullYear(),
+            });
+        }
+    },
+
```
[the remark this answers is not in the document]
Comment (lukebarnard1) [reply src/DateUtils.js]: oh cool, I'll do that instead

**Comment by lukebarnard1** [comment]
>lots of stuff in this PR

Should I split it up into several commits? I realise that it's all been done in one blob but need not go down in history as such.

**Comment by lukebarnard1** [comment]
@richvdh test flake: https://travis-ci.org/matrix-org/matrix-react-sdk/builds/270466858?utm_source=github_status&utm_medium=notification

**Pull request merged**